drivers: timer: Harmonize mtime-based RISC-V timers

VynDragon opened the pull request:
This commit replaces a bunch of ifdefs and bindings with a single extensible binding, and makes all standard mtime system timers consistent and the same.

Reason for removing scr_machine_timer special divider case is the lack of usage, and I believe that is something that should be in the clock_control driver or similar, not here.

Follows discussions in : https://github.com/zephyrproject-rtos/zephyr/pull/84103

This has only been tested in practice with the hardware available to me, which means a unsupported yet platform, I encourage maintainers to check it works as expected on their boards.

VynDragon committed:
drivers: timer: Harmonize mtime-based RISC-V timers

This commit replaces a bunch of ifdefs and bindings with a single
extensible binding, and makes all standard mtime system timers consistent.

Signed-off-by: Camille BAUD <mail@massdriver.space>

andyross commented on drivers/timer/riscv_machine_timer.c:
```diff
@@ -99,11 +45,11 @@
+#define CYCLES_MAX_2 ((uint64_t)CYCLE_DIFF_MAX)
+#define CYCLES_MAX_3 MIN(CYCLES_MAX_1, CYCLES_MAX_2)
+#define CYCLES_MAX_4 (CYCLES_MAX_3 / 2 + CYCLES_MAX_3 / 4)
+#define CYCLES_MAX   (CYCLES_MAX_4 + LSB_GET(CYCLES_MAX_4))
```
Unrelated whitespace change should probably be in a separate commit.

VynDragon replied:
This was the output of clang-format, but yea, probably, will update soon.

VynDragon replied:
Done.

andyross commented on drivers/timer/riscv_machine_timer.c:
```diff
@@ -138,14 +84,6 @@ static void set_mtimecmp(uint64_t time)
-#ifdef MTIMER_HAS_DIVIDER
-	*(volatile uint32_t *)MTIMEDIV_REG =
-		CONFIG_RISCV_MACHINE_TIMER_SYSTEM_CLOCK_DIVIDER;
-#endif
```
Is this being removed because it's dead code?  If so, in a separate commit from the DTS changes please.

VynDragon replied:
Yes, as far as I can tell it is dead code,  and there should be a more generic implementation of the ability to set the divider via clock drivers (clock_set_rate #72102 ) put in this driver later (I am aware that for at least the platforms i work on, there is a settable multiplier that sets the mtime rate from main clock). will do.

VynDragon replied:
Done.

andyross reviewed (dismissed):
This version I like very much, thanks!  A few nitpicks, but IMHO this is how DTS-configured drivers are supposed to look.

VynDragon committed:
drivers: timer: Remove unused divider setter in RISC-V machine timer

This removes a unused function for a unused binding,
and is prelude to a refactoring

Signed-off-by: Camille BAUD <mail@massdriver.space>

VynDragon committed:
drivers: timer: run clang-format on RISC-V machine timer file

Clang-Format changes whitespacing and longer lines are now allowed

Signed-off-by: Camille BAUD <mail@massdriver.space>

nandojve commented on dts/riscv/renode_riscv32_virt.dtsi:
```diff
@@ -50,6 +50,13 @@
+		mtimer: timer@200bff8 {
+			compatible = "riscv,mtime-timer";
+			interrupts-extended = <&hlic 7>;
+			reg = <0x200BFF8 0x8 0x2004000 0x8>;
```
If I'm not mistaken, you need to use lower case on all hex entries

```diff
-0x200BFF8
+0x200bff8
```

VynDragon replied:
Done.

nandojve commented on dts/bindings/timer/riscv,mtime-timer.yaml:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2024 MASSDRIVER EI (massdriver.space)
```
2024-2025 ?

VynDragon replied:
New year is hard lol. Done.

nandojve reviewed (changes requested):
@VynDragon ,

Just cosmetics.

@andyross ,

Maybe a topic for another PR: should this driver be initialized as device instead sysinit?

nandojve commented:
Hi @soburi ,

Is #54919 broke timer implementation for nuclei once it dropped CONFIG_RISCV_MACHINE_TIMER_SYSTEM_CLOCK_DIVIDER?

As @VynDragon is fixing this driver I was wondering if the `clk-divider` property on `nuclei,systimer.yaml` is supposed to be used as the CONFIG_RISCV_MACHINE_TIMER_SYSTEM_CLOCK_DIVIDER. If is that the case, maybe we could move it to

soburi commented:
@nandojve 

> As @VynDragon is fixing this driver I was wondering if the `clk-divider` property on `nuclei,systimer.yaml` is supposed to be used as the CONFIG_RISCV_MACHINE_TIMER_SYSTEM_CLOCK_DIVIDER. If is that the case, maybe we could move it to `riscv,mtime-timer`, right?

I haven't worked on this code for a while, so there are some parts I can't remember right away, but the problem with GD32V is using mtimer, but which is implemented uniquely, and the system clock is divided by 4 to use as the timer clock. Therefore, it is necessary to calculate the divider amount *4, /4 to convert between the system clock and the timer clock. Once this calculation is included, the property itself can be moved.
I made it a somewhat generic property, considering the possibility that there might be other similar implementations, but if you don't need to assume that implementations other than GD32V use the divider, I think it would be fine to eliminate the property and incorporate the calculation fixedly.

VynDragon commented:
@soburi This seem to be handled in the same way we handle it with bflb, by setting the value manually in the SoC Kconfigs to the value it's supposed to be: https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/gd/gd32/gd32vf103/Kconfig.defconfig.gd32vf103

Additionally, I have a couple questions for reviewers to improve/fix this PR:

- Is MSIP (interrupt 3) supposed to trigger a timer update? If so, I need to update this PR to add 3 where I didnt add it in the DTS entry.
- sifive,clint0 binding is now mostly unused. I believe it should be renamed to riscv,clint. But should it be kept and its clic counterpart added, to be later used instead of RISCV_HAS_CLIC and RISCV_HAS_CLINT, sourcing this kind of infos from DTS instead of Kconfig, or should it be removed entirely and replaced with Kconfig?

fkokosinski commented on dts/bindings/timer/riscv,mtime-timer.yaml:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2025 MASSDRIVER EI (massdriver.space)
+# SPDX-License-Identifier: Apache-2.0
+
+description: RISC-V Standard mtime Timer.
```
```suggestion
description: RISC-V Machine Timer.
```

The RISC-V spec uses this name to refer to that block, so let's reflect that here as well

VynDragon replied:
Done.

fkokosinski commented on dts/bindings/timer/riscv,mtime-timer.yaml:
```diff
@@ -0,0 +1,14 @@
+
+description: RISC-V Standard mtime Timer.
+
+compatible: "riscv,mtime-timer"
```
```suggestion
compatible: "riscv,machine-timer"
```

VynDragon replied:
Done.

fkokosinski reviewed (changes requested):
Thanks a lot for this cleanup!

VynDragon commented:
> * Is MSIP (interrupt 3) supposed to trigger a timer update? If so, I need to update this PR to add 3 where I didnt add it in the DTS entry.
> 
> * sifive,clint0 binding is now mostly unused. I believe it should be renamed to riscv,clint. But should it be kept and its clic counterpart added, to be later used instead of RISCV_HAS_CLIC and RISCV_HAS_CLINT, sourcing this kind of infos from DTS instead of Kconfig, or should it be removed entirely and replaced with Kconfig?

I'd like those questions answered before it is merged to avoid any mistakes, please. In theory the PR is good as is, but it can be further cleaned up and avoid a possible issue later.
